Source/Shared/arcana/threading/cancellation.h

```diff
@@ -2,28 +2,118 @@
 
 #include "arcana/containers/ticketed_collection.h"
 
+#include <gsl/gsl>
+
 #include <algorithm>
-#include <cassert>
 #include <atomic>
-#include <memory>
-#include <vector>
-
+#include <cassert>
 #include <functional>
+#include <memory>
+#include <optional>
 #include <vector>
 
 namespace arcana
 {
-    class cancellation
+    namespace internal
     {
-        using collection = ticketed_collection<std::function<void>>;
+        class cancellation_impl
+        {
+            using collection = ticketed_collection<std::function<void()>>;
+
+        public:
+            using ticket = collection::ticket;
+            using ticket_scope = collection::ticket_scope;
+
+            template<typename CallableT>
+            ticket add_listener(CallableT&& callback, std::function<void()>& copied)
+            {
+                std::scoped_lock<std::mutex> lock{ m_mutex };
```

Comment:
Why this change? Isn't scoped_lock heavier than lock_guard, and intended for cases where you are trying to lock on multiple mutexes?

Reply:
I'm not aware that scoped_lock is heavier than lock_guard. The only reasoning for using lock_guard I could find is here: https://stackoverflow.com/questions/43019598/stdlock-guard-or-stdscoped-lock?msclkid=6aad9a81bc4111eca785163544f6d052